ui/src/functions/api.js

```diff
@@ -0,0 +1,67 @@
+// Determine the URL of an API route by inspecting the URL of the current page
+// and assuming the API URL is /auth/api/v1 at the same host.
+function apiUrl(route) {
+  if (typeof window === 'undefined') return;
+  return new URL(`/auth/api/v1${route}`, window.location.href).href;
+}
+
+// Redirect to the /login route and request a return to the current page.
+function apiLoginRedirect() {
```

Comment:
The case I was trying to handle here is when the user is successfully logged in and browsing the site, and then their authentication cookies expire. I'm not sure how that works with a user profile context; we currently only call the login route once. The above pattern seems equivalent to putting the authentication handling in the `LoginContext`, but then everything starts failing with errors instead of a redirect when the user's credentials expire.